repeater/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.repeater/PayloadParser.scala

```diff
@@ -0,0 +1,107 @@
+  * into event fully however it can be converted mostly. In order to
+  * get more information, look at the comments of "ReconstructedEvent"
+  */
+object PayloadParser {
+  def parse(payload: JsonObject): Either[ParsingError, ReconstructedEvent] = {
+    val jsonsWithFlattenSchemaKey = getJsonsWithFlattenSchema(payload)
+
+    val modified = payload
+      .add("contexts", Json.obj())
+      .add("derived_contexts", Json.obj())
+
+    modified.asJson.as[Event] match {
+      case Left(e) =>
+        Left(
+          ParsingError(
+            payload.asJson.noSpaces,
+            NonEmptyList.of(EventParsingError(e.message, e.history.foldLeft(List[String]())((l, i) => i.toString::l)))
+          )
+        )
+      case Right(e) =>
+        Right(ReconstructedEvent(e, jsonsWithFlattenSchemaKey))
+    }
+
+  }
+
+  private def getJsonsWithFlattenSchema(payload: JsonObject): List[JsonWithFlattenSchemaKey] = {
+    val jsonPrefixes = List("contexts", "unstruct_event")
+    jsonPrefixes.flatMap(getJsonsWithPrefix(payload, _))
+  }
+
+  private def getJsonsWithPrefix(payload: JsonObject, prefix: String): List[JsonWithFlattenSchemaKey] =
+    payload.filter {
+      case (key, value) => key.startsWith(prefix) && value != Json.obj() && value != Json.Null
+    }
+    .toList.map {
+      case (flattenSchemaKey, data) => JsonWithFlattenSchemaKey(flattenSchemaKey, data)
+    }
+
+  /**
+    * Represents event which is reconstructed from payload of the failed insert
+    * It consists of two part because schema keys of the self describing JSONs
+    * such as unstruct_event, contexts or derived_contexts are flatten in the
+    * payload. For example schema key with com.snowplowanalytics.snowplow as
+    * vendor, web_page as name, 1.0.0 as version is converted to
+    * "contexts_com_snowplowanalytics_snowplow_web_page_1_0_0" and this key could
+    * not be reconverted to its original schema key deterministically. Therefore,
+    * they can not be behaved like self describing JSONs after reconstruction because
+    * their schema key are unknown. Therefore, reconstructed event split into two part
+    * as atomic and JSONs with schema key
+    * @param atomic event instance which only consists of atomic event parts
+    * @param jsonsWithFlattenSchemaKey list of JsonWithFlattenSchemaKey instances
+    *        they consist of reconstructed self describing JSONs from the payload
+    */
+  final case class ReconstructedEvent(atomic: Event, jsonsWithFlattenSchemaKey: List[JsonWithFlattenSchemaKey])
+
+  implicit val reconstructedEventEncoder: Encoder[ReconstructedEvent] =
+    Encoder.instance {
+      case ReconstructedEvent(atomic, jsonsWithFlattenSchemaKey) =>
+        Json.obj(
+          "atomic" := atomic.asJson,
+          "jsonsWithFlattenSchemaKey" := jsonsWithFlattenSchemaKey.asJson
+        )
+    }
+
+  //TODO Enes: Ask Anton if we can give jsons of unstruct_event, contexts, derived_contexts
```

Comment:
I call them "shredded types" usually. "Self-describing entities" is also feasible option.